# Notebook 05: MLOps - Pipeline de produccion

## Objetivo

Documentar y demostrar como el modelo de lead scoring pasa de un notebook experimental 
a un sistema desplegable en produccion. Este notebook cubre dos entornos diferenciados:

### Dos entornos: Demo y Produccion

| | **DEMO (Streamlit Cloud)** | **PRODUCCION (propuesta)** |
|---|---|---|
| **Que es** | App interactiva ya desplegada | Arquitectura profesional propuesta |
| **URL** | [raona-lead-scoring.streamlit.app](https://raona-lead-scoring.streamlit.app) | `localhost:8000` (Docker local) |
| **Scoring** | Formulario individual + carga de CSV | API REST tiempo real + batch via Airflow |
| **Orquestacion** | Manual (el usuario sube datos) | Airflow DAGs automaticos (semanal/mensual) |
| **Monitorizacion** | No | PSI mensual + reentrenamiento condicional |
| **Estado** | **Desplegado y funcional** | **Codigo generado, no desplegado** |

Ambos entornos comparten los mismos artefactos del modelo (`lead_scorer.pkl`, `preprocessor.pkl`, 
`clustering.pkl`, `feature_names.pkl`) y la misma logica de scoring.

| Seccion | Contenido | Entorno |
|---------|----------|---------|
| **5.1** | Pipeline end-to-end: de datos crudos a prediccion | Comun |
| **5.2** | Tracking de experimentos con MLflow | Comun |
| **5.3** | Servir el modelo: API FastAPI + Docker | Produccion |
| **5.4** | Demo: Streamlit Cloud | Demo |
| **5.5** | Pipelines de produccion (archivos .py) | Produccion |
| **5.6** | DAG 1 - Scoring pipeline (frecuencia alta) | Produccion |
| **5.7** | DAG 2 - Monitoring + Retraining (frecuencia baja) | Produccion |
| **5.8** | Arquitectura completa | Comun |
| **5.9** | Trabajo futuro y roadmap | Comun |

## Por que MLOps?

Un modelo en un notebook no es un producto. Para que Raona pueda usar el lead scoring 
en su dia a dia, necesita:
- Un **API** que reciba datos de un contacto y devuelva un score
- Un **contenedor** que se pueda desplegar en cualquier servidor
- **Orquestacion** para procesar nuevos contactos automaticamente
- **Monitorizacion** para saber cuando el modelo deja de funcionar bien
- **Reproducibilidad** para poder reentrenar con datos nuevos

## Imports y configuración

In [1]:
import pandas as pd
import numpy as np
import os
import pickle
import json
import warnings
warnings.filterwarnings('ignore', category=FutureWarning)
warnings.filterwarnings('ignore', category=UserWarning)

from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import average_precision_score, roc_auc_score

import lightgbm as lgb
import mlflow
import mlflow.sklearn

import plotly.express as px
import plotly.graph_objects as go
import plotly.io as pio
pio.templates.default = 'plotly_white'

# Rutas
PROJECT_ROOT = os.path.abspath(os.path.join(os.getcwd(), '..'))
WORKING_DATA = os.path.join(PROJECT_ROOT, '..', '_working', 'data')
WORKING_MODELS = os.path.join(PROJECT_ROOT, '..', '_working', 'models')
DELIVERABLE_MODELS = os.path.join(PROJECT_ROOT, 'app', 'models')
MLRUNS_DIR = os.path.join(PROJECT_ROOT, '..', '_working', 'mlruns')
APP_DIR = os.path.join(PROJECT_ROOT, 'app')

SEED = 42
np.random.seed(SEED)

print('Configuracion lista')
print(f'Modelos en: {DELIVERABLE_MODELS}')

/Users/acaballito/Library/Python/3.9/lib/python/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


Configuracion lista
Modelos en: /Users/acaballito/Library/CloudStorage/GoogleDrive-adriana.caballero@gmail.com/.shortcut-targets-by-id/1LzrxzfxIAZukyDLfOvioF7Z2ESoOXZGz/TFM/acaballero/TFM_deliverables/app/models


---
## 5.1 Pipeline end-to-end

Demostramos que el pipeline completo funciona: desde un diccionario JSON con los datos 
de un contacto hasta la prediccion del score, cluster y recomendacion.

Esto es exactamente lo que hace tanto la **demo Streamlit** como la **API de produccion** 
propuesta: la misma funcion `score_lead()` encapsula la logica de scoring comun a ambos entornos.

In [2]:
# Cargar artefactos del modelo (produccion)
with open(os.path.join(DELIVERABLE_MODELS, 'preprocessor.pkl'), 'rb') as f:
    preprocessor = pickle.load(f)

with open(os.path.join(DELIVERABLE_MODELS, 'lead_scorer.pkl'), 'rb') as f:
    lead_scorer = pickle.load(f)

with open(os.path.join(DELIVERABLE_MODELS, 'clustering.pkl'), 'rb') as f:
    clustering_bundle = pickle.load(f)

with open(os.path.join(DELIVERABLE_MODELS, 'feature_names.pkl'), 'rb') as f:
    feature_names = pickle.load(f)

# Feature list del modelo
if isinstance(feature_names, list):
    FEATURE_COLS = feature_names
else:
    FEATURE_COLS = feature_names

print(f'Artefactos cargados:')
print(f'  Preprocessor: {type(preprocessor).__name__}')
print(f'  Lead scorer: {type(lead_scorer).__name__}')
print(f'  Features: {len(FEATURE_COLS)}')
print(f'  Clustering: {list(clustering_bundle.keys())}')

Artefactos cargados:
  Preprocessor: Pipeline
  Lead scorer: LGBMClassifier
  Features: 39
  Clustering: ['kmeans', 'imputer', 'scaler', 'features']


In [3]:
def score_lead(contact_data: dict) -> dict:
    """Pipeline completo: datos de contacto -> score + cluster + recomendacion."""
    # 1. Crear DataFrame
    df_input = pd.DataFrame([contact_data])
    for col in FEATURE_COLS:
        if col not in df_input.columns:
            df_input[col] = np.nan
    
    # 2. Preprocesar
    X = preprocessor.transform(df_input[FEATURE_COLS])
    
    # 3. Lead score
    score = lead_scorer.predict_proba(X)[:, 1][0]
    
    # 4. Cluster
    cluster_features = clustering_bundle['features']
    df_cluster = df_input[cluster_features].copy()
    X_cluster = clustering_bundle['scaler'].transform(
        clustering_bundle['imputer'].transform(df_cluster)
    )
    cluster = int(clustering_bundle['kmeans'].predict(X_cluster)[0])
    
    # 5. Clasificar riesgo
    if score >= 0.5:
        risk_level = 'ALTO'
    elif score >= 0.2:
        risk_level = 'MEDIO'
    else:
        risk_level = 'BAJO'
    
    return {
        'lead_score': round(float(score), 4),
        'cluster': cluster,
        'risk_level': risk_level,
        'recommended_channel': 'LinkedIn',  # Mayor reply rate: 5.0% vs 2.9%
        'recommended_day': 'Jueves',  # Mayor reply rate: 8.7%
    }

print('Funcion score_lead() definida')

Funcion score_lead() definida


In [4]:
# Demo: scorear un contacto real del pool de no contactados
pool_path = os.path.join(WORKING_DATA, 'not_contacted_pool_scored.parquet')
df_pool = pd.read_parquet(pool_path)

# Tomar el contacto con mayor score como ejemplo
top_contact = df_pool.nlargest(1, 'lead_score').iloc[0]
print(f'=== Demo: contacto real del pool (top 1 por lead_score) ===')
print(f'  Empresa: {top_contact.get("Company name", "N/A")}')
print(f'  Seniority: {top_contact.get("ai_SENIORITY", "N/A")}')
print(f'  Departamento: {top_contact.get("ai_DEPARTMENT", "N/A")}')
print()

# Scorear usando la funcion score_lead
demo_features = {f: top_contact[f] for f in FEATURE_COLS if f in top_contact.index}
result = score_lead(demo_features)
print('=== Resultado del scoring ===')
for k, v in result.items():
    print(f'  {k}: {v}')

=== Demo: contacto real del pool (top 1 por lead_score) ===
  Empresa: Real Madrid C.F.
  Seniority: MANAGER
  Departamento: Sales & MKT



=== Resultado del scoring ===
  lead_score: 0.0844
  cluster: 1
  risk_level: BAJO
  recommended_channel: LinkedIn
  recommended_day: Jueves


In [5]:
# Verificar con datos reales del dataset
df = pd.read_parquet(os.path.join(WORKING_DATA, 'predictions.parquet'))

pos_sample = df[df['target_replied'] == 1].iloc[0]
neg_sample = df[df['target_replied'] == 0].iloc[0]

for label, sample in [('POSITIVO (respondio)', pos_sample), ('NEGATIVO (no respondio)', neg_sample)]:
    contact = {col: sample[col] for col in FEATURE_COLS if col in sample.index}
    result = score_lead(contact)
    print(f'\n=== Contacto {label} ===')
    print(f'  Score: {result["lead_score"]}')
    print(f'  Risk level: {result["risk_level"]}')
    print(f'  Cluster: {result["cluster"]}')


=== Contacto POSITIVO (respondio) ===
  Score: 0.8972
  Risk level: ALTO
  Cluster: 2

=== Contacto NEGATIVO (no respondio) ===
  Score: 0.0552
  Risk level: BAJO
  Cluster: 0


---
## 5.2 Tracking de experimentos con MLflow

MLflow permite registrar y comparar todos los experimentos realizados en NB04. 
Registramos retroactivamente los resultados de la competicion de modelos.

En producción, cada reentrenamiento del DAG 2 se registraria automáticamente en MLflow, 
creando un historial completo de versiones del modelo.

In [6]:
# Configurar MLflow con almacenamiento local
mlflow.set_tracking_uri(f'file://{os.path.abspath(MLRUNS_DIR)}')
experiment_name = 'raona_lead_scoring'
mlflow.set_experiment(experiment_name)

print(f'MLflow tracking URI: {mlflow.get_tracking_uri()}')
print(f'Experimento: {experiment_name}')

MLflow tracking URI: file:///Users/acaballito/Library/CloudStorage/GoogleDrive-adriana.caballero@gmail.com/.shortcut-targets-by-id/1LzrxzfxIAZukyDLfOvioF7Z2ESoOXZGz/TFM/acaballero/_working/mlruns
Experimento: raona_lead_scoring


In [7]:
# Cargar metricas reales de NB04
import json as json_lib
metrics_path = os.path.join(WORKING_DATA, 'metrics.json')
with open(metrics_path) as f:
    nb04_metrics = json_lib.load(f)

model_m = nb04_metrics['model']

# Registrar el experimento de NB04 retroactivamente
experiment = {
    'name': 'LightGBM_PRODUCTION',
    'params': {'model_type': 'LightGBM', 'tuned': True, 'n_trials': 50,
               'n_features': nb04_metrics['n_features']},
    'metrics': {'pr_auc': model_m['PR-AUC'], 'roc_auc': model_m['ROC-AUC'],
                'precision_at_100': model_m['Precision@100'], 'lift_10pct': model_m['Lift@10%']},
}

with mlflow.start_run(run_name=experiment['name']):
    mlflow.log_params(experiment['params'])
    mlflow.log_metrics(experiment['metrics'])
    mlflow.sklearn.log_model(lead_scorer, 'lead_scorer')

print(f'  Registrado: {experiment["name"]} (PR-AUC={experiment["metrics"]["pr_auc"]:.3f})')
print(f'\nTotal runs registrados: 1')

2026/03/19 19:10:55 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.


2026/03/19 19:10:57 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


  Registrado: LightGBM_PRODUCTION (PR-AUC=0.303)

Total runs registrados: 1


In [8]:
# Visualizar el experimento registrado
exp_df = pd.DataFrame([{
    'Modelo': experiment['name'],
    'PR-AUC': experiment['metrics']['pr_auc'],
    'ROC-AUC': experiment['metrics']['roc_auc']
}])

fig = go.Figure()
fig.add_trace(go.Bar(name='PR-AUC', x=exp_df['Modelo'], y=exp_df['PR-AUC'],
                     marker_color='#2ecc71', text=exp_df['PR-AUC'].round(3), textposition='auto'))
fig.add_trace(go.Bar(name='ROC-AUC', x=exp_df['Modelo'], y=exp_df['ROC-AUC'],
                     marker_color='#3498db', text=exp_df['ROC-AUC'].round(3), textposition='auto'))
fig.update_layout(title='MLflow: Experimento registrado', barmode='group', height=400)
fig.show()

---
## 5.3 Infraestructura de produccion

El sistema de produccion se compone de dos bloques funcionales, desplegados en 
seis servicios Docker orquestados por `docker-compose.yml`:

### Arquitectura de containers

| Bloque | Servicio | Rol | Imagen | Puerto |
|--------|----------|-----|--------|--------|
| **API** | `lead-scorer-api` | Scoring REST en tiempo real | `Dockerfile` (Python + FastAPI) | 8000 |
| **Airflow** | `airflow-webserver` | Interfaz web de Airflow | `Dockerfile.airflow` | 8080 |
| | `airflow-scheduler` | Ejecuta los DAGs programados | `Dockerfile.airflow` | - |
| | `airflow-triggerer` | Triggers asincronos | `Dockerfile.airflow` | - |
| | `airflow-init` | Inicializa BD y usuario admin | `Dockerfile.airflow` | - |
| | `postgres` | Metadata DB de Airflow | `postgres:13` (oficial) | 5432 |

### Por que dos Dockerfiles

Cada bloque tiene dependencias diferentes:

- **`Dockerfile`** (API): imagen ligera con FastAPI, uvicorn y el modelo. Solo sirve 
  predicciones. No necesita Airflow ni su ecosistema.
- **`Dockerfile.airflow`**: parte de la imagen oficial de Airflow y le anade `libgomp1` 
  (OpenMP, requerido por LightGBM) y las dependencias de ML (`scikit-learn==1.6.1`, 
  `lightgbm==4.6.0`, `pyarrow`). Ejecuta los pipelines de datos.

Separarlos mantiene cada imagen ligera y con responsabilidad unica.

### Por que Airflow se despliega en servicios separados

En desarrollo, Airflow puede correr como un solo proceso (`airflow standalone`). 
En produccion, se separa en servicios independientes por tres razones:

1. **Resiliencia** -- Si el webserver se cae, el scheduler sigue ejecutando los DAGs. 
   Los pipelines no se detienen porque la interfaz falle.
2. **Escalabilidad** -- Cada servicio consume recursos diferentes. El scheduler necesita 
   CPU para orquestar; el webserver necesita memoria para servir la UI. Separados, se 
   pueden escalar independientemente.
3. **Mantenimiento** -- Se puede actualizar el webserver sin detener la ejecucion de los 
   DAGs. En un entorno productivo, esto significa zero downtime en los pipelines.

### Dependencias separadas

El proyecto mantiene dos archivos de dependencias para evitar conflictos:

| Archivo | Entorno | Contenido |
|---------|---------|-----------|
| `requirements-api.txt` | API Docker (produccion) | FastAPI, uvicorn, sklearn, lightgbm, pydantic |
| `requirements.txt` | Streamlit Cloud (demo) | Streamlit, sklearn, lightgbm, plotly, openpyxl |

Las versiones de `scikit-learn` y `lightgbm` deben coincidir con las usadas para generar 
los artefactos `.pkl` (`sklearn==1.6.1`, `lightgbm==4.6.0`).

### API de scoring

El **Container 1 (API)** esta siempre corriendo y atiende peticiones de scoring en tiempo real:

```
POST /score  ->  FastAPI  ->  modelo  ->  JSON response
GET /health  ->  verificacion de que el servicio esta activo
```

Las celdas siguientes generan los archivos de infraestructura directamente desde el notebook.

In [9]:
# Generar api.py
api_code = '''"""FastAPI endpoint para el lead scoring de Raona.

Uso:
    uvicorn api:app --host 0.0.0.0 --port 8000

Documentacion interactiva:
    http://localhost:8000/docs
"""
import os
import pickle
import numpy as np
import pandas as pd
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel, Field
from typing import Optional

# --- Cargar modelo al iniciar ---
MODEL_DIR = os.path.join(os.path.dirname(__file__), "models")

with open(os.path.join(MODEL_DIR, "preprocessor.pkl"), "rb") as f:
    preprocessor = pickle.load(f)
with open(os.path.join(MODEL_DIR, "lead_scorer.pkl"), "rb") as f:
    lead_scorer = pickle.load(f)
with open(os.path.join(MODEL_DIR, "clustering.pkl"), "rb") as f:
    clustering_bundle = pickle.load(f)
with open(os.path.join(MODEL_DIR, "feature_names.pkl"), "rb") as f:
    feature_names = pickle.load(f)

FEATURE_COLS = feature_names if isinstance(feature_names, list) else feature_names
CLUSTER_FEATURES = clustering_bundle["features"]


# --- Schemas ---
class ContactInput(BaseModel):
    """Datos de entrada del contacto."""
    years_in_company: Optional[float] = Field(None, description="Anos en la empresa")
    number_of_connections: Optional[float] = Field(None, description="Conexiones en LinkedIn")
    number_of_employees: Optional[float] = Field(None, description="Empleados de la empresa")
    year_founded: Optional[float] = Field(None, description="Ano de fundacion")
    hiring_on_linkedin: Optional[float] = Field(None, description="1 si tiene ofertas activas")
    six_months_growth: Optional[float] = Field(None, description="Crecimiento plantilla 6 meses")
    two_years_growth: Optional[float] = Field(None, description="Crecimiento plantilla 2 anos")
    yearly_growth: Optional[float] = Field(None, description="Crecimiento plantilla anual")
    fe_seniority_ord: Optional[float] = Field(None, description="Seniority ordinal (0-5)")
    fe_type_of_contact_ord: Optional[float] = Field(None, description="Tipo contacto ordinal (0-5)")
    fe_fit_approved: Optional[float] = Field(None, description="FIT aprobado (0/1)")
    fe_fit_data_approved: Optional[float] = Field(None, description="FIT DATA aprobado (0/1)")
    fe_company_age: Optional[float] = Field(None, description="Edad de la empresa")
    fe_log_employees: Optional[float] = Field(None, description="log1p(empleados)")
    fe_company_size_bucket: Optional[float] = Field(None, description="Bucket tamano (0-4)")
    fe_log_connections: Optional[float] = Field(None, description="log1p(conexiones)")
    fe_headcount_momentum: Optional[float] = Field(None, description="Momentum crecimiento")
    fe_has_bio: Optional[float] = Field(None, description="Tiene bio en LinkedIn (0/1)")
    fe_microsoft_flag: Optional[float] = Field(None, description="Usa Microsoft (0/1)")
    fe_department_encoded: Optional[float] = Field(None, description="Departamento target-encoded")
    ext_ms_maturity_score: Optional[float] = Field(None, description="Score madurez Microsoft")
    ext_has_competitor_tech: Optional[float] = Field(None, description="Usa tech competidora (0/1)")
    nlp_report_length: Optional[float] = Field(None, description="Longitud company report")
    nlp_contact_report_length: Optional[float] = Field(None, description="Longitud contact report AI")
    nlp_has_momentum: Optional[float] = Field(None, description="Tiene info momentum (0/1)")
    nlp_urgency_score: Optional[float] = Field(None, description="Score de urgencia")
    nlp_embedding_01: Optional[float] = Field(None, description="Embedding UMAP dim 1")
    nlp_embedding_02: Optional[float] = Field(None, description="Embedding UMAP dim 2")
    nlp_embedding_03: Optional[float] = Field(None, description="Embedding UMAP dim 3")
    nlp_topic: Optional[float] = Field(None, description="Topic cluster asignado")


class ScoreOutput(BaseModel):
    """Resultado del scoring."""
    lead_score: float = Field(description="Probabilidad de respuesta (0-1)")
    cluster: int = Field(description="Segmento asignado (0-6)")
    risk_level: str = Field(description="ALTO (>0.5), MEDIO (0.2-0.5), BAJO (<0.2)")
    recommended_channel: str = Field(description="Canal recomendado")
    recommended_day: str = Field(description="Mejor dia para contactar")


# --- App ---
app = FastAPI(
    title="Raona Lead Scoring API",
    description="Scoring de leads B2B con modelo LightGBM",
    version="2.0.0",
)


@app.get("/health")
def health_check():
    return {
        "status": "ok",
        "model": "lead_scorer",
        "n_features": len(FEATURE_COLS),
        "n_clusters": clustering_bundle["kmeans"].n_clusters,
    }


@app.post("/score", response_model=ScoreOutput)
def score_contact(contact: ContactInput):
    try:
        field_mapping = {
            "years_in_company": "Years in company",
            "number_of_connections": "Number of connections",
            "number_of_employees": "Number of employees",
            "year_founded": "Year founded",
            "hiring_on_linkedin": "Hiring on LinkedIn",
            "six_months_growth": "Six months headcount growth",
            "two_years_growth": "Two years headcount growth",
            "yearly_growth": "Yearly headcount growth",
        }
        contact_dict = {}
        for field_name, value in contact.dict().items():
            col_name = field_mapping.get(field_name, field_name)
            contact_dict[col_name] = value

        df_input = pd.DataFrame([contact_dict])
        for col in FEATURE_COLS:
            if col not in df_input.columns:
                df_input[col] = np.nan

        X = preprocessor.transform(df_input[FEATURE_COLS])
        score = float(lead_scorer.predict_proba(X)[:, 1][0])

        df_cluster = df_input[CLUSTER_FEATURES].copy()
        for col in CLUSTER_FEATURES:
            if col not in df_cluster.columns:
                df_cluster[col] = np.nan
        X_cluster = clustering_bundle["scaler"].transform(
            clustering_bundle["imputer"].transform(df_cluster)
        )
        cluster = int(clustering_bundle["kmeans"].predict(X_cluster)[0])

        risk_level = "ALTO" if score >= 0.5 else "MEDIO" if score >= 0.2 else "BAJO"

        return ScoreOutput(
            lead_score=round(score, 4), cluster=cluster, risk_level=risk_level,
            recommended_channel="LinkedIn", recommended_day="Jueves",
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
'''

api_path = os.path.join(APP_DIR, 'api.py')
with open(api_path, 'w') as f:
    f.write(api_code)

print(f'API generado: {api_path}')
print(f'Modelo: lead_scorer.pkl ({len(FEATURE_COLS)} features)')

API generado: /Users/acaballito/Library/CloudStorage/GoogleDrive-adriana.caballero@gmail.com/.shortcut-targets-by-id/1LzrxzfxIAZukyDLfOvioF7Z2ESoOXZGz/TFM/acaballero/TFM_deliverables/app/api.py
Modelo: lead_scorer.pkl (39 features)


#### Nota sobre el diseño de la API

La API actual es un **endpoint de scoring** que recibe features pre-procesadas 
(e.g., `fe_seniority_ord`, `nlp_embedding_01`). En un despliegue real, la cadena completa seria:

1. **Datos crudos** (perfil LinkedIn, tecnologias, informes AI) llegan via Airflow/ingestion
2. **Pipeline de feature engineering** (equivalente a NB03) transforma los datos crudos
3. **API de scoring** (este endpoint) recibe el vector de features y devuelve el score

Esta separación de responsabilidades es intencional: el feature engineering se ejecuta en batch 
(Airflow), mientras que el scoring es en tiempo real (API). La API no necesita conocer 
la logica de transformación, solo el formato de entrada del modelo.


In [10]:
# Simular una llamada POST /score
print('=== Simulacion de llamada POST /score ===')
print('\nRequest body (JSON):')
request_body = {
    'number_of_connections': 1200,
    'number_of_employees': 800,
    'fe_seniority_ord': 3,
    'fe_type_of_contact_ord': 4,
    'fe_company_size_bucket': 3,
    'fe_microsoft_flag': 1,
}
print(json.dumps(request_body, indent=2))

response = score_lead(request_body)
print('\nResponse:')
print(json.dumps(response, indent=2))

=== Simulacion de llamada POST /score ===

Request body (JSON):
{
  "number_of_connections": 1200,
  "number_of_employees": 800,
  "fe_seniority_ord": 3,
  "fe_type_of_contact_ord": 4,
  "fe_company_size_bucket": 3,
  "fe_microsoft_flag": 1
}

Response:
{
  "lead_score": 0.4087,
  "cluster": 0,
  "risk_level": "MEDIO",
  "recommended_channel": "LinkedIn",
  "recommended_day": "Jueves"
}


In [11]:
# Generar Dockerfile para Container API
dockerfile = '''FROM python:3.9-slim

WORKDIR /app

RUN apt-get update && apt-get install -y --no-install-recommends \\
    libgomp1 curl \\
    && rm -rf /var/lib/apt/lists/*

COPY requirements-api.txt .
RUN pip install --no-cache-dir -r requirements-api.txt

COPY api.py .
COPY models/lead_scorer.pkl models/
COPY models/preprocessor.pkl models/
COPY models/clustering.pkl models/
COPY models/feature_names.pkl models/

EXPOSE 8000

HEALTHCHECK --interval=30s --timeout=5s --retries=3 \\
    CMD curl -f http://localhost:8000/health || exit 1

CMD ["uvicorn", "api:app", "--host", "0.0.0.0", "--port", "8000"]
'''

with open(os.path.join(APP_DIR, 'Dockerfile'), 'w') as f:
    f.write(dockerfile)
print('Dockerfile generado (Container API: Python + FastAPI + curl)')

Dockerfile generado (Container API: Python + FastAPI + curl)


In [12]:
# Generar docker-compose.yml (orquestacion completa: API + Airflow + Postgres)
compose = '''x-airflow-common:
  &airflow-common
  build:
    context: .
    dockerfile: Dockerfile.airflow
  environment:
    &airflow-common-env
    AIRFLOW__CORE__EXECUTOR: LocalExecutor
    AIRFLOW__CORE__SQL_ALCHEMY_CONN: postgresql+psycopg2://airflow:airflow@postgres/airflow
    AIRFLOW__CORE__FERNET_KEY: ''
    AIRFLOW__CORE__DAGS_ARE_PAUSED_AT_CREATION: 'false'
    AIRFLOW__CORE__LOAD_EXAMPLES: 'false'
    AIRFLOW__API__AUTH_BACKEND: 'airflow.api.auth.backend.basic_auth'
    LEAD_SCORING_BASE_DIR: /opt/airflow/data
    LEAD_SCORING_MODEL_DIR: /opt/airflow/models
    PIPELINES_DIR: /opt/airflow/dags
  volumes:
    - ./dags:/opt/airflow/dags
    - ./pipelines:/opt/airflow/dags/pipelines
    - ./models:/opt/airflow/models
    - ../../_working/data:/opt/airflow/data
    - ./logs:/opt/airflow/logs
  user: "${AIRFLOW_UID:-50000}:0"
  depends_on:
    &airflow-common-depends-on
    postgres:
      condition: service_healthy

services:
  lead-scorer-api:
    build: .
    ports:
      - "8000:8000"
    restart: unless-stopped
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:8000/health"]
      interval: 30s
      timeout: 5s
      retries: 3

  postgres:
    image: postgres:13
    environment:
      POSTGRES_USER: airflow
      POSTGRES_PASSWORD: airflow
      POSTGRES_DB: airflow
    ports:
      - "5432:5432"
    volumes:
      - postgres-db-volume:/var/lib/postgresql/data
    healthcheck:
      test: ["CMD", "pg_isready", "-U", "airflow"]
      interval: 5s
      retries: 5
    restart: always

  airflow-init:
    <<: *airflow-common
    entrypoint: /bin/bash
    command:
      - -c
      - |
        mkdir -p /sources/logs /sources/dags /sources/plugins
        exec /entrypoint airflow version
    environment:
      <<: *airflow-common-env
      _AIRFLOW_DB_MIGRATE: 'true'
      _AIRFLOW_WWW_USER_CREATE: 'true'
      _AIRFLOW_WWW_USER_USERNAME: airflow
      _AIRFLOW_WWW_USER_PASSWORD: airflow
      _PIP_ADDITIONAL_REQUIREMENTS: ''
    user: "0:0"
    volumes:
      - .:/sources

  airflow-webserver:
    <<: *airflow-common
    command: webserver
    ports:
      - "8080:8080"
    healthcheck:
      test: ["CMD", "curl", "--fail", "http://localhost:8080/health"]
      interval: 10s
      timeout: 10s
      retries: 5
    restart: always
    depends_on:
      <<: *airflow-common-depends-on
      airflow-init:
        condition: service_completed_successfully

  airflow-scheduler:
    <<: *airflow-common
    command: scheduler
    healthcheck:
      test: ["CMD-SHELL", 'airflow jobs check --job-type SchedulerJob --hostname "$${HOSTNAME}"']
      interval: 10s
      timeout: 10s
      retries: 5
    restart: always
    depends_on:
      <<: *airflow-common-depends-on
      airflow-init:
        condition: service_completed_successfully

  airflow-triggerer:
    <<: *airflow-common
    command: triggerer
    healthcheck:
      test: ["CMD-SHELL", 'airflow jobs check --job-type TriggererJob --hostname "$${HOSTNAME}"']
      interval: 10s
      timeout: 10s
      retries: 5
    restart: always
    depends_on:
      <<: *airflow-common-depends-on
      airflow-init:
        condition: service_completed_successfully

volumes:
  postgres-db-volume:
'''

with open(os.path.join(APP_DIR, 'docker-compose.yml'), 'w') as f:
    f.write(compose)

# Generar requirements-api.txt (dependencias del API Docker, NO de Streamlit)
requirements_api = '''fastapi==0.104.1
uvicorn==0.24.0
pandas>=2.0.0
numpy>=1.24.0,<2.1.0
scikit-learn==1.6.1
lightgbm==4.6.0
pydantic==2.5.2
'''

with open(os.path.join(APP_DIR, 'requirements-api.txt'), 'w') as f:
    f.write(requirements_api)

print('docker-compose.yml generado (6 servicios: API + Postgres + 4 Airflow)')
print('requirements-api.txt generado (dependencias API produccion)')
print('\n=== Desplegar ===')
print('  cd app/')
print('  docker compose up -d')
print('  # API en http://localhost:8000')
print('  # Airflow en http://localhost:8080 (airflow/airflow)')

docker-compose.yml generado (6 servicios: API + Postgres + 4 Airflow)
requirements-api.txt generado (dependencias API produccion)

=== Desplegar ===
  cd app/
  docker compose up -d
  # API en http://localhost:8000
  # Airflow en http://localhost:8080 (airflow/airflow)


In [13]:
# Generar Dockerfile.airflow (imagen custom para los containers de Airflow)
# Necesario porque LightGBM requiere libgomp1 (OpenMP) que no viene en la imagen base
dockerfile_airflow = '''FROM apache/airflow:2.10.4-python3.9

USER root
RUN apt-get update && apt-get install -y --no-install-recommends \\
    libgomp1 \\
    && rm -rf /var/lib/apt/lists/*

USER airflow
RUN pip install --no-cache-dir \\
    scikit-learn==1.6.1 lightgbm==4.6.0 pyarrow
'''

with open(os.path.join(APP_DIR, 'Dockerfile.airflow'), 'w') as f:
    f.write(dockerfile_airflow)
print('Dockerfile.airflow generado (Airflow + libgomp1 + sklearn + lightgbm)')

Dockerfile.airflow generado (Airflow + libgomp1 + sklearn + lightgbm)


---
## 5.4 Demo: Streamlit Cloud

La demo desplegada es una **app interactiva en Streamlit Cloud** que permite al equipo 
comercial de Raona explorar los resultados del modelo sin necesidad de infraestructura propia.

**URL:** [raona-lead-scoring.streamlit.app](https://raona-lead-scoring.streamlit.app)

### Estructura de la app (3 pestanas)

| Pestana | Funcion | Datos que usa |
|---------|---------|---------------|
| **Dashboard** | KPIs globales, distribucion de scores, analisis por canal/timing, segmentacion por cluster | `sample_contacts.parquet`, `daily_analytics_ES.parquet` |
| **Lead Scorer** | Formulario para scorear un contacto individual con recomendacion de outreach | `models/*.pkl` (scoring en tiempo real) |
| **Batch Scorer** | Subir CSV de contactos, scorear todos a la vez, descargar ranking en Excel | `models/*.pkl` + CSV del usuario |

### Archivos de la demo

| Archivo | Funcion |
|---------|---------|
| `app/streamlit_app.py` | Codigo principal (~1000 lineas: CSS personalizado + logica + visualizaciones Plotly) |
| `app/requirements.txt` | Dependencias de Streamlit Cloud (`scikit-learn==1.6.1`, `lightgbm==4.6.0`) |
| `app/data/sample_contacts.parquet` | Dataset de contactos con scores pre-calculados para el dashboard |
| `app/data/daily_analytics_ES.parquet` | Metricas diarias agregadas para graficos temporales |
| `app/data/not_contacted_pool.csv` | Pool de 4,735 contactos no contactados para batch scoring |
| `app/models/*.pkl` | Artefactos del modelo (los mismos que usaria la API de produccion) |

### Despliegue

Streamlit Cloud conecta directamente al repositorio GitHub (`acaballito/raona-lead-scoring`). 
Cada push a la rama `main` dispara una reconstruccion automatica de la app. No requiere 
Docker, servidor propio ni configuracion de infraestructura.

> **Nota:** El archivo `streamlit_app.py` no se genera desde este notebook por su extension 
> (~1000 lineas con CSS personalizado y visualizaciones interactivas). El codigo fuente 
> completo esta disponible en `app/streamlit_app.py`.

### Diferencia clave con produccion

La demo y la propuesta de produccion comparten los mismos artefactos `.pkl`, pero difieren en:
- **Dependencias separadas**: `requirements.txt` (Streamlit) vs `requirements-api.txt` (API Docker)
- **Scoring**: La demo usa la logica directamente en Python; la produccion la expone via API REST
- **Orquestacion**: La demo es manual; la produccion usa Airflow para batch scoring semanal

---
## 5.5 Pipelines de produccion (archivos .py)

Los DAGs de Airflow (secciones 5.6 y 5.7) invocan modulos Python que encapsulan la logica 
de cada notebook. Cada modulo es una extraccion directa del codigo de los notebooks, 
adaptada para ejecutarse de forma autonoma en un entorno de produccion.

| Modulo | Notebook origen | Funcion |
|--------|----------------|---------|
| `pipelines/ingest.py` | NB01 | Carga CSV de Enginy, valida esquema, crea target, guarda Parquet |
| `pipelines/transform.py` | NB03 | Feature engineering completo (39 features) |
| `pipelines/score.py` | NB04 | Scoring con `lead_scorer.pkl` + clustering + priorizacion |
| `pipelines/monitor.py` | NB05 | Calcula PSI por feature para detectar drift |
| `pipelines/retrain.py` | NB05 | Reentrena LightGBM con datos acumulados |
| `pipelines/validate.py` | NB05 | Compara candidato vs actual, promueve si es mejor |

Las celdas siguientes generan estos archivos en `app/pipelines/` directamente desde el notebook, 
siguiendo el mismo patron que `api.py` y `Dockerfile` en la seccion 5.3.

In [14]:
# Generar pipelines/__init__.py
PIPELINES_DIR = os.path.join(APP_DIR, 'pipelines')
os.makedirs(PIPELINES_DIR, exist_ok=True)

init_code = '"""Pipeline modules para Raona Lead Scoring MLOps.\n\nExtraido de los notebooks para uso en produccion con Airflow.\n"""\n'

with open(os.path.join(PIPELINES_DIR, '__init__.py'), 'w') as f:
    f.write(init_code)
print('pipelines/__init__.py generado')

pipelines/__init__.py generado


In [15]:
# Generar pipelines/ingest.py (logica de NB01)
ingest_code = '''"""Ingest: carga y validacion de datos nuevos.

Logica extraida de NB01. Carga un CSV de contactos exportado de Enginy,
valida el esquema y guarda en formato Parquet.
"""
import os
import re
import pandas as pd
import numpy as np
import logging

logger = logging.getLogger(__name__)

REQUIRED_COLS = [
    "LinkedIn profile ID", "Company name", "Campaign engagement status",
    "Job title", "Number of employees", "Industry",
]


def load_contacts(csv_path: str) -> pd.DataFrame:
    """Carga contacts CSV y aplica limpieza basica."""
    logger.info(f"Cargando {csv_path}")
    df = pd.read_csv(csv_path, encoding="utf-8-sig", low_memory=False)
    logger.info(f"Filas cargadas: {len(df):,}")
    return df


def validate_schema(df: pd.DataFrame) -> bool:
    """Valida que el DataFrame tenga las columnas minimas requeridas."""
    missing = [c for c in REQUIRED_COLS if c not in df.columns]
    if missing:
        logger.error(f"Columnas faltantes: {missing}")
        return False
    logger.info("Esquema validado correctamente")
    return True


def create_target(df: pd.DataFrame) -> pd.DataFrame:
    """Crea variable objetivo target_replied desde Campaign engagement status."""
    df["target_replied"] = df["Campaign engagement status"].str.contains(
        "Replied", case=False, na=False
    ).astype(int)
    logger.info(f"Target creado: {df['target_replied'].sum()} positivos ({df['target_replied'].mean():.1%})")
    return df


def filter_valid_contacts(df: pd.DataFrame) -> pd.DataFrame:
    """Filtra contactos sin mensajes enviados y no validos."""
    n_before = len(df)
    status = df["Campaign engagement status"]
    contacted = status.str.contains("Sent|Replied|Connection Accepted", case=False, na=False)
    df = df[contacted].copy()
    if "Microsoft?" in df.columns:
        df = df[df["Microsoft?"] != -1].copy()
    if "FIT" in df.columns:
        fit_no = df["FIT"].str.strip().str.upper() == "NO"
        df = df[~fit_no].copy()
    logger.info(f"Contactos filtrados: {n_before:,} -> {len(df):,}")
    return df


def extract_reply_message_number(df: pd.DataFrame) -> pd.DataFrame:
    """Extrae el numero de mensaje de respuesta del status."""
    def _extract(status):
        if pd.isna(status):
            return np.nan
        match = re.search(r"Replied\\s*\\((\\d+)\\)", status)
        return int(match.group(1)) if match else np.nan
    df["reply_message_number"] = df["Campaign engagement status"].apply(_extract)
    return df


def run(csv_path: str, output_path: str) -> str:
    """Pipeline completo de ingestion."""
    df = load_contacts(csv_path)
    if not validate_schema(df):
        raise ValueError("Esquema de datos invalido")
    df = create_target(df)
    df = filter_valid_contacts(df)
    df = extract_reply_message_number(df)
    df.to_parquet(output_path, index=False)
    logger.info(f"Guardado: {output_path} ({len(df):,} filas)")
    return output_path
'''

with open(os.path.join(PIPELINES_DIR, 'ingest.py'), 'w') as f:
    f.write(ingest_code)
print('pipelines/ingest.py generado (logica de NB01)')

pipelines/ingest.py generado (logica de NB01)


In [16]:
# Generar pipelines/transform.py (logica de NB03, CORREGIDO con 39 features)
# Requiere: scikit-learn==1.6.1, lightgbm==4.6.0
transform_code = '''"""Transform: feature engineering para datos nuevos.

Logica extraida de NB03. Aplica las mismas transformaciones que el notebook
a datos nuevos para que sean compatibles con el modelo (39 features).

Requiere: scikit-learn==1.6.1, lightgbm==4.6.0
"""
import re
import numpy as np
import pandas as pd
import logging

logger = logging.getLogger(__name__)

SENIORITY_MAP = {"CLEVEL": 4, "DIRECTOR": 3, "MANAGER": 2, "LEAD": 1, "JR": 0}

TYPE_OF_CONTACT_MAP = {
    "KEY_DECISION_MAKER": 5, "BUYER_CHAMPION": 4, "CHAMPION": 3,
    "INFLUENCER": 3, "REFERRAL": 2, "NULL": 0,
}

URGENCY_KEYWORDS = [
    "crecimiento", "expansion", "transformacion digital",
    "nuevo proyecto", "inversion", "presupuesto",
    "contratando", "hiring", "growth",
    "licitacion", "concurso",
    "migracion", "modernizacion",
    "urgente", "inmediato", "prioridad",
]

HR_KEYWORDS = ["HR", "RRHH", "PEOPLE", "RECURSOS HUMANOS", "HUMAN RESOURCES", "TALENT"]
COMMS_KEYWORDS = ["COMUNICACION", "COMMUNICATIONS", "MARKETING", "BRAND"]

# Scores de madurez Microsoft (extraido de NB03)
TECH_SCORES = {
    # Microsoft ecosystem (positivo)
    "azure": 3, "microsoft azure": 3,
    "dynamics 365": 3, "dynamics": 3,
    "microsoft 365": 2, "office 365": 2, "microsoft office": 2,
    "power bi": 2, "power platform": 2, "power apps": 2, "power automate": 2,
    "sharepoint": 2,
    "teams": 1, "microsoft teams": 1,
    "active directory": 1, "entra id": 1, "azure ad": 1,
    "exchange": 1, "exchange online": 1,
    "windows server": 1, "sql server": 1,
    "intune": 2, "endpoint manager": 2,
    "copilot": 2,
    # Competidores (negativo)
    "google workspace": -1, "google cloud": -1, "gcp": -1,
    "aws": -1, "amazon web services": -1,
    "salesforce": -1, "hubspot": -1,
    "slack": -1, "zoom": -1,
}

# Mapeo de productos Raona a keywords tecnologicas (extraido de NB03)
PRODUCT_TECH_MAP = {
    "comunica": ["teams", "microsoft teams", "skype"],
    "colabora": ["sharepoint", "onedrive", "power apps", "power automate", "power platform"],
    "infra": ["azure", "microsoft azure", "windows server", "sql server"],
    "ia": ["copilot", "azure ai", "cognitive", "openai"],
    "data": ["power bi", "fabric", "synapse", "sql server"],
    "workplace": ["microsoft 365", "office 365", "intune", "endpoint manager"],
    "maite": ["entra id", "azure ad", "active directory", "purview", "intune"],
}


def clean_type_of_contact(val):
    """Limpia el campo TYPE OF CONTACT a categorias estandar."""
    if pd.isna(val):
        return "NULL"
    val_clean = re.sub(r"[^a-zA-Z\\s-]", "", str(val)).strip().upper()
    if "DECISOR" in val_clean:
        return "KEY_DECISION_MAKER"
    if "CHAMPION" in val_clean and "BUYER" in val_clean:
        return "BUYER_CHAMPION"
    if "CHAMPION" in val_clean:
        return "CHAMPION"
    if "INFLUENCER" in val_clean:
        return "INFLUENCER"
    if "REFERER" in val_clean or "REFERIDOR" in val_clean:
        return "REFERRAL"
    return "NULL"


def correct_department(row):
    """Corrige departamento usando Job title."""
    dept = row.get("ai_DEPARTMENT", np.nan)
    title = str(row.get("Job title", "")).upper()
    if any(kw in title for kw in HR_KEYWORDS):
        return "HR"
    if any(kw in title for kw in COMMS_KEYWORDS):
        return "COMMUNICATIONS"
    return dept


def count_urgency_keywords(text):
    """Cuenta palabras clave de urgencia en texto."""
    if pd.isna(text):
        return 0
    text_lower = str(text).lower()
    return sum(1 for kw in URGENCY_KEYWORDS if kw in text_lower)


def score_tech_stack(tech_string):
    """Calcula score de madurez Microsoft y presencia de competidores."""
    if pd.isna(tech_string):
        return 0, 0
    tech_lower = str(tech_string).lower()
    ms_score = 0
    has_competitor = 0
    for tech, score in TECH_SCORES.items():
        if tech in tech_lower:
            if score > 0:
                ms_score += score
            else:
                has_competitor = 1
    return ms_score, has_competitor


def tech_contains(tech_str, keywords):
    """Verifica si el string de tecnologias contiene alguna keyword."""
    if pd.isna(tech_str):
        return 0
    tech_lower = str(tech_str).lower()
    return int(any(kw in tech_lower for kw in keywords))


def run(df: pd.DataFrame, global_mean: float = 0.06) -> pd.DataFrame:
    """Aplica feature engineering completo a un DataFrame (39 features)."""
    logger.info(f"Feature engineering en {len(df):,} filas")

    # --- Codificaciones ordinales ---
    if "ai_SENIORITY" in df.columns:
        df["fe_seniority_ord"] = df["ai_SENIORITY"].map(SENIORITY_MAP)

    if "ai_TYPE_OF_CONTACT" in df.columns:
        df["ai_TYPE_OF_CONTACT_clean"] = df["ai_TYPE_OF_CONTACT"].apply(clean_type_of_contact)
        df["fe_type_of_contact_ord"] = df["ai_TYPE_OF_CONTACT_clean"].map(TYPE_OF_CONTACT_MAP)

    # --- FIT ---
    if "ai_FIT" in df.columns:
        df["fe_fit_approved"] = df["ai_FIT"].apply(
            lambda x: 1.0 if pd.notna(x) and "APROBADO" in str(x).upper()
                       and "DESAPROBADO" not in str(x).upper() else 0.0
        )
    if "ai_FIT_DATA" in df.columns:
        df["fe_fit_data_approved"] = df["ai_FIT_DATA"].map({"SI": 1, "NO": 0, "COMPETITOR": 0, "DUDA": 0})

    # --- Numericas ---
    if "Year founded" in df.columns:
        df["fe_company_age"] = 2026 - df["Year founded"]
    if "Number of employees" in df.columns:
        df["fe_log_employees"] = np.log1p(df["Number of employees"])
        bins = [0, 10, 50, 250, 1000, np.inf]
        labels = [0, 1, 2, 3, 4]
        df["fe_company_size_bucket"] = pd.cut(
            df["Number of employees"], bins=bins, labels=labels, right=False
        ).astype(float)
    if "Number of connections" in df.columns:
        df["fe_log_connections"] = np.log1p(df["Number of connections"])

    # --- Headcount momentum ---
    for col in ["Six months headcount growth", "Yearly headcount growth", "Two years headcount growth"]:
        if col not in df.columns:
            df[col] = 0.0
    df["fe_headcount_momentum"] = (
        0.5 * df["Six months headcount growth"].fillna(0) +
        0.3 * df["Yearly headcount growth"].fillna(0) +
        0.2 * df["Two years headcount growth"].fillna(0)
    )

    # --- Binarias ---
    if "Professional email" in df.columns:
        df["fe_has_email"] = df["Professional email"].notna().astype(int)
    if "Profile bio" in df.columns:
        df["fe_has_bio"] = df["Profile bio"].notna().astype(int)
    if "ai_Microsoft" in df.columns:
        df["fe_microsoft_flag"] = (df["ai_Microsoft"] == 1).astype(int)

    # --- Department encoding ---
    if "ai_DEPARTMENT" in df.columns:
        df["fe_department_corrected"] = df.apply(correct_department, axis=1)
        df["fe_department_encoded"] = global_mean  # Fallback con media global

    # --- Enrichment externo (ext_) ---
    if "Technologies used" in df.columns:
        tech_results = df["Technologies used"].apply(score_tech_stack)
        df["ext_ms_maturity_score"] = tech_results.apply(lambda x: x[0])
        df["ext_has_competitor_tech"] = tech_results.apply(lambda x: x[1])
    else:
        df["ext_ms_maturity_score"] = 0
        df["ext_has_competitor_tech"] = 0

    # --- Tech fit por producto (fe_tech_fit_*) ---
    if "Technologies used" in df.columns:
        for product, keywords in PRODUCT_TECH_MAP.items():
            col_name = f"fe_tech_fit_{product}"
            df[col_name] = df["Technologies used"].apply(lambda x, kw=keywords: tech_contains(x, kw))
    else:
        for product in PRODUCT_TECH_MAP:
            df[f"fe_tech_fit_{product}"] = 0

    # --- NLP basicas ---
    if "ai_COMPANY_REPORT" in df.columns:
        df["nlp_report_length"] = df["ai_COMPANY_REPORT"].str.len().fillna(0)
    else:
        df["nlp_report_length"] = 0
    if "ai_CONTACT_REPORT" in df.columns:
        df["nlp_contact_report_length"] = df["ai_CONTACT_REPORT"].str.len().fillna(0)
    else:
        df["nlp_contact_report_length"] = 0
    if "ai_MOMENTUM" in df.columns:
        df["nlp_has_momentum"] = df["ai_MOMENTUM"].notna().astype(int)
        urgency_m = df["ai_MOMENTUM"].apply(count_urgency_keywords)
        urgency_r = df.get("ai_COMPANY_REPORT", pd.Series(dtype=str)).apply(count_urgency_keywords)
        df["nlp_urgency_score"] = pd.concat([urgency_m, urgency_r], axis=1).max(axis=1)
    else:
        df["nlp_has_momentum"] = 0
        df["nlp_urgency_score"] = 0

    # --- NLP embeddings (requeririan modelo de embeddings en produccion) ---
    for col in ["nlp_embedding_01", "nlp_embedding_02", "nlp_embedding_03", "nlp_topic"]:
        if col not in df.columns:
            df[col] = 0.0

    n_features = len([c for c in df.columns if c.startswith(("fe_", "nlp_", "ext_"))])
    logger.info(f"Features creadas: {n_features} (fe_ + nlp_ + ext_)")
    return df
'''

with open(os.path.join(PIPELINES_DIR, 'transform.py'), 'w') as f:
    f.write(transform_code)
print('pipelines/transform.py generado (logica de NB03, 39 features)')
print('  Incluye: TECH_SCORES, PRODUCT_TECH_MAP, score_tech_stack(), tech_contains()')

pipelines/transform.py generado (logica de NB03, 39 features)
  Incluye: TECH_SCORES, PRODUCT_TECH_MAP, score_tech_stack(), tech_contains()


In [17]:
# Generar pipelines/score.py (logica de NB04)
score_code = '''"""Score: aplica modelo de lead scoring y clustering a datos nuevos.

Logica extraida de NB04. Usa el modelo LightGBM para scoring de contactos nuevos.

Requiere: scikit-learn==1.6.1, lightgbm==4.6.0
"""
import os
import pickle
import numpy as np
import pandas as pd
import logging

logger = logging.getLogger(__name__)


def load_artifacts(model_dir: str) -> dict:
    """Carga todos los artefactos del modelo."""
    artifacts = {}
    for name in ["lead_scorer", "preprocessor", "clustering", "feature_names"]:
        path = os.path.join(model_dir, f"{name}.pkl")
        if os.path.exists(path):
            with open(path, "rb") as f:
                artifacts[name] = pickle.load(f)
            logger.info(f"Cargado: {name}")
    return artifacts


def score_leads(df: pd.DataFrame, artifacts: dict) -> pd.DataFrame:
    """Aplica scoring y asigna clusters."""
    features = artifacts["feature_names"]
    prep = artifacts["preprocessor"]
    model = artifacts["lead_scorer"]

    X = df.reindex(columns=features, fill_value=np.nan)
    X_processed = prep.transform(X)
    df["lead_score"] = model.predict_proba(X_processed)[:, 1]
    logger.info(f"Scores generados: media={df['lead_score'].mean():.3f}")

    clustering = artifacts["clustering"]
    cluster_feats = clustering["features"]
    X_cl = df.reindex(columns=cluster_feats, fill_value=0)
    X_cl_scaled = clustering["scaler"].transform(
        clustering["imputer"].transform(X_cl)
    )
    df["cluster"] = clustering["kmeans"].predict(X_cl_scaled)

    df["priority"] = pd.cut(
        df["lead_score"], bins=[-0.01, 0.1, 0.3, 1.01],
        labels=["Low", "Medium", "High"],
    )
    logger.info(f"Distribucion: {df['priority'].value_counts().to_dict()}")
    return df


def run(parquet_path: str, model_dir: str, output_path: str) -> str:
    """Pipeline de scoring."""
    df = pd.read_parquet(parquet_path)
    artifacts = load_artifacts(model_dir)
    df = score_leads(df, artifacts)
    df.to_parquet(output_path, index=False)
    logger.info(f"Guardado: {output_path} ({len(df):,} filas)")
    return output_path
'''

# Generar pipelines/monitor.py (logica de NB05)
monitor_code = '''"""Monitor: detecta drift en features usando PSI.

Logica extraida de NB05. Calcula Population Stability Index entre
los datos de entrenamiento y los datos nuevos.
"""
import numpy as np
import pandas as pd
import logging

logger = logging.getLogger(__name__)


def calculate_psi(expected: np.ndarray, actual: np.ndarray, bins: int = 10) -> float:
    """Calcula Population Stability Index entre dos distribuciones."""
    breakpoints = np.percentile(expected[~np.isnan(expected)], np.linspace(0, 100, bins + 1))
    breakpoints = np.unique(breakpoints)
    if len(breakpoints) < 3:
        return 0.0
    expected_counts = np.histogram(expected[~np.isnan(expected)], bins=breakpoints)[0]
    actual_counts = np.histogram(actual[~np.isnan(actual)], bins=breakpoints)[0]
    expected_pct = (expected_counts + 1) / (expected_counts.sum() + len(expected_counts))
    actual_pct = (actual_counts + 1) / (actual_counts.sum() + len(actual_counts))
    psi = np.sum((actual_pct - expected_pct) * np.log(actual_pct / expected_pct))
    return float(psi)


def classify_psi(psi: float) -> str:
    """Clasifica el nivel de drift segun el PSI."""
    if psi < 0.1:
        return "OK"
    elif psi < 0.25:
        return "MONITORIZAR"
    return "ALERTA"


def run(train_path: str, new_path: str, features: list) -> pd.DataFrame:
    """Calcula PSI para cada feature entre datos de entrenamiento y nuevos."""
    train_df = pd.read_parquet(train_path)
    new_df = pd.read_parquet(new_path)
    results = []
    for feat in features:
        if feat in train_df.columns and feat in new_df.columns:
            train_vals = train_df[feat].values.astype(float)
            new_vals = new_df[feat].values.astype(float)
            psi = calculate_psi(train_vals, new_vals)
            status = classify_psi(psi)
            results.append({"feature": feat, "psi": round(psi, 4), "status": status})
            if status == "ALERTA":
                logger.warning(f"DRIFT DETECTADO en {feat}: PSI={psi:.4f}")
    psi_df = pd.DataFrame(results)
    logger.info(f"PSI calculado para {len(results)} features. Alertas: {(psi_df['status'] == 'ALERTA').sum()}")
    return psi_df
'''

# Generar pipelines/retrain.py (logica de NB05)
retrain_code = '''"""Retrain: reentrenamiento del modelo con datos acumulados.

Solo se ejecuta si el monitor detecta drift significativo.

Requiere: scikit-learn==1.6.1, lightgbm==4.6.0
"""
import os
import pickle
import numpy as np
import pandas as pd
import logging

from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import average_precision_score

logger = logging.getLogger(__name__)
SEED = 42


def run(data_path: str, model_dir: str, feature_names_path: str) -> dict:
    """Reentrena el modelo con los datos acumulados."""
    try:
        import lightgbm as lgb
    except ImportError:
        logger.error("LightGBM no disponible")
        return {"status": "error", "message": "LightGBM not installed"}

    df = pd.read_parquet(data_path)
    logger.info(f"Datos para reentrenamiento: {len(df):,} filas")

    with open(feature_names_path, "rb") as f:
        features = pickle.load(f)

    X = df[features].copy()
    y = df["target_replied"].copy()

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=SEED, stratify=y
    )

    preprocessor = Pipeline([
        ("imputer", SimpleImputer(strategy="median")),
        ("scaler", StandardScaler()),
    ])
    X_train_p = preprocessor.fit_transform(X_train)
    X_test_p = preprocessor.transform(X_test)

    neg_pos_ratio = (y_train == 0).sum() / max((y_train == 1).sum(), 1)
    model = lgb.LGBMClassifier(
        n_estimators=300, learning_rate=0.05, max_depth=6,
        num_leaves=31, scale_pos_weight=neg_pos_ratio,
        random_state=SEED, verbose=-1,
    )
    model.fit(X_train_p, y_train)

    y_proba = model.predict_proba(X_test_p)[:, 1]
    pr_auc = average_precision_score(y_test, y_proba)
    logger.info(f"Nuevo modelo PR-AUC: {pr_auc:.4f}")

    candidate_dir = os.path.join(model_dir, "candidate")
    os.makedirs(candidate_dir, exist_ok=True)
    with open(os.path.join(candidate_dir, "lead_scorer.pkl"), "wb") as f:
        pickle.dump(model, f)
    with open(os.path.join(candidate_dir, "preprocessor.pkl"), "wb") as f:
        pickle.dump(preprocessor, f)

    return {
        "status": "success", "pr_auc": pr_auc,
        "train_size": len(X_train), "test_size": len(X_test),
        "candidate_dir": candidate_dir,
    }
'''

# Generar pipelines/validate.py (logica de NB05)
validate_code = '''"""Validate: compara modelo candidato vs modelo actual.

Promueve el nuevo modelo a produccion solo si mejora el PR-AUC
en un holdout set independiente.

Requiere: scikit-learn==1.6.1, lightgbm==4.6.0
"""
import os
import pickle
import shutil
import numpy as np
import pandas as pd
import logging

from sklearn.metrics import average_precision_score

logger = logging.getLogger(__name__)
MIN_IMPROVEMENT = 0.01


def run(data_path: str, model_dir: str, candidate_dir: str, feature_names_path: str) -> dict:
    """Compara candidato vs actual y promueve si es mejor."""
    df = pd.read_parquet(data_path)
    y = df["target_replied"]

    with open(feature_names_path, "rb") as f:
        features = pickle.load(f)
    X = df[features].copy()

    # Modelo actual
    with open(os.path.join(model_dir, "preprocessor.pkl"), "rb") as f:
        current_prep = pickle.load(f)
    with open(os.path.join(model_dir, "lead_scorer.pkl"), "rb") as f:
        current_model = pickle.load(f)
    X_current = current_prep.transform(X)
    current_score = average_precision_score(y, current_model.predict_proba(X_current)[:, 1])

    # Modelo candidato
    with open(os.path.join(candidate_dir, "preprocessor.pkl"), "rb") as f:
        candidate_prep = pickle.load(f)
    with open(os.path.join(candidate_dir, "lead_scorer.pkl"), "rb") as f:
        candidate_model = pickle.load(f)
    X_candidate = candidate_prep.transform(X)
    candidate_score = average_precision_score(y, candidate_model.predict_proba(X_candidate)[:, 1])

    improvement = candidate_score - current_score
    logger.info(f"PR-AUC actual: {current_score:.4f}, candidato: {candidate_score:.4f}, mejora: {improvement:+.4f}")

    result = {
        "current_pr_auc": current_score,
        "candidate_pr_auc": candidate_score,
        "improvement": improvement,
    }

    if improvement >= MIN_IMPROVEMENT:
        shutil.copy(os.path.join(candidate_dir, "lead_scorer.pkl"), os.path.join(model_dir, "lead_scorer.pkl"))
        shutil.copy(os.path.join(candidate_dir, "preprocessor.pkl"), os.path.join(model_dir, "preprocessor.pkl"))
        result["promoted"] = True
        logger.info("Modelo candidato PROMOVIDO a produccion")
    else:
        result["promoted"] = False
        logger.info(f"Candidato NO promovido (mejora {improvement:.4f} < umbral {MIN_IMPROVEMENT})")

    shutil.rmtree(candidate_dir, ignore_errors=True)
    return result
'''

# Escribir todos los archivos
for name, code in [('score.py', score_code), ('monitor.py', monitor_code),
                    ('retrain.py', retrain_code), ('validate.py', validate_code)]:
    with open(os.path.join(PIPELINES_DIR, name), 'w') as f:
        f.write(code)
    print(f'pipelines/{name} generado')

# Verificacion
print('\n=== Pipelines generados ===')
for name in ['__init__.py', 'ingest.py', 'transform.py', 'score.py',
             'monitor.py', 'retrain.py', 'validate.py']:
    path = os.path.join(PIPELINES_DIR, name)
    exists = 'OK' if os.path.exists(path) else 'FALTA'
    size = os.path.getsize(path) if os.path.exists(path) else 0
    print(f'  [{exists}] {name} ({size:,} bytes)')

pipelines/score.py generado


pipelines/monitor.py generado
pipelines/retrain.py generado
pipelines/validate.py generado

=== Pipelines generados ===
  [OK] __init__.py (118 bytes)
  [OK] ingest.py (2,919 bytes)
  [OK] transform.py (8,842 bytes)
  [OK] score.py (2,096 bytes)
  [OK] monitor.py (2,161 bytes)
  [OK] retrain.py (2,395 bytes)
  [OK] validate.py (2,409 bytes)


---
## 5.6 DAG 1: Scoring pipeline (frecuencia alta)

Este DAG se ejecuta **cada vez que Raona exporta contactos nuevos de Enginy** 
(tipicamente semanal). Su unica funcion es procesar y scorear el batch de contactos.

### Flujo

```
ingest ──> transform ──> score ──> done
```

| Paso | Modulo | Que hace |
|------|--------|----------|
| **ingest** | `pipelines/ingest.py` | Carga CSV nuevo de Enginy, valida esquema, guarda Parquet |
| **transform** | `pipelines/transform.py` | Aplica feature engineering (mismas transformaciones que NB03) |
| **score** | `pipelines/score.py` | Scorea todos los contactos con `lead_scorer.pkl`, asigna clusters |

### Resultado

Un archivo CSV (o tabla en base de datos) con todos los contactos nuevos ordenados 
por `lead_score` descendente, con su cluster, nivel de riesgo y recomendaciones.

### Frecuencia

Semanal o bajo demanda. Se puede disparar manualmente o programar en Airflow.

El siguiente codigo genera `app/dags/scoring_dag.py`:

In [18]:
# Generar scoring_dag.py (DAG 1: scoring semanal)
DAGS_DIR = os.path.join(APP_DIR, 'dags')
os.makedirs(DAGS_DIR, exist_ok=True)

scoring_dag_code = '''"""Airflow DAG 1: Scoring semanal de contactos nuevos.

Pipeline: ingest >> transform >> score >> done

Se ejecuta semanalmente cuando Raona exporta nuevos contactos de Enginy.
En demo mode, divide el dataset existente 80/20 para simular datos nuevos.
"""
from datetime import datetime, timedelta
from airflow import DAG
from airflow.operators.python import PythonOperator
from airflow.operators.empty import EmptyOperator

import os
import sys
import pandas as pd
import logging

logger = logging.getLogger(__name__)

BASE_DIR = os.environ.get("LEAD_SCORING_BASE_DIR", "/opt/airflow/data")
MODEL_DIR = os.environ.get("LEAD_SCORING_MODEL_DIR", "/opt/airflow/models")
RAW_DATA_DIR = os.path.join(BASE_DIR, "raw")
PROCESSED_DIR = os.path.join(BASE_DIR, "processed")
SCORED_DIR = os.path.join(BASE_DIR, "scored")

PIPELINES_DIR = os.environ.get("PIPELINES_DIR", "/opt/airflow/dags")
if PIPELINES_DIR not in sys.path:
    sys.path.insert(0, PIPELINES_DIR)

default_args = {
    "owner": "raona-data",
    "depends_on_past": False,
    "email_on_failure": False,
    "retries": 1,
    "retry_delay": timedelta(minutes=5),
}


def _setup_demo_mode(**kwargs):
    """En demo mode, divide el dataset existente 80/20."""
    from sklearn.model_selection import train_test_split
    os.makedirs(RAW_DATA_DIR, exist_ok=True)
    os.makedirs(PROCESSED_DIR, exist_ok=True)
    os.makedirs(SCORED_DIR, exist_ok=True)
    full_data_path = os.path.join(BASE_DIR, "modeling_dataset_final.parquet")
    if not os.path.exists(full_data_path):
        logger.warning(f"Dataset no encontrado: {full_data_path}")
        return
    df = pd.read_parquet(full_data_path)
    train, new = train_test_split(df, test_size=0.2, random_state=42, stratify=df["target_replied"])
    train.to_parquet(os.path.join(PROCESSED_DIR, "historical.parquet"), index=False)
    new.to_parquet(os.path.join(RAW_DATA_DIR, "new_contacts.parquet"), index=False)
    logger.info(f"Demo: {len(train)} historicos, {len(new)} nuevos")


def _ingest(**kwargs):
    """Ejecuta pipeline de ingestion."""
    input_path = os.path.join(RAW_DATA_DIR, "new_contacts.parquet")
    output_path = os.path.join(PROCESSED_DIR, "ingested.parquet")
    if input_path.endswith(".parquet"):
        df = pd.read_parquet(input_path)
        df.to_parquet(output_path, index=False)
    else:
        from pipelines import ingest
        ingest.run(input_path, output_path)
    return output_path


def _transform(**kwargs):
    """Ejecuta feature engineering."""
    from pipelines import transform
    input_path = os.path.join(PROCESSED_DIR, "ingested.parquet")
    df = pd.read_parquet(input_path)
    df = transform.run(df)
    output_path = os.path.join(PROCESSED_DIR, "transformed.parquet")
    df.to_parquet(output_path, index=False)
    return output_path


def _score(**kwargs):
    """Aplica scoring."""
    from pipelines import score
    input_path = os.path.join(PROCESSED_DIR, "transformed.parquet")
    output_path = os.path.join(SCORED_DIR, f"scored_{datetime.now().strftime('%Y%m%d')}.parquet")
    score.run(input_path, MODEL_DIR, output_path)
    return output_path


with DAG(
    dag_id="raona_scoring_weekly",
    default_args=default_args,
    description="Scoring semanal: ingest -> transform -> score",
    schedule_interval="@weekly",
    start_date=datetime(2026, 1, 1),
    catchup=False,
    tags=["raona", "ml", "scoring"],
) as dag:

    setup = PythonOperator(task_id="setup_demo", python_callable=_setup_demo_mode)
    ingest = PythonOperator(task_id="ingest", python_callable=_ingest)
    transform = PythonOperator(task_id="transform", python_callable=_transform)
    score = PythonOperator(task_id="score", python_callable=_score)
    done = EmptyOperator(task_id="done")

    setup >> ingest >> transform >> score >> done
'''

with open(os.path.join(DAGS_DIR, 'scoring_dag.py'), 'w') as f:
    f.write(scoring_dag_code)
print('dags/scoring_dag.py generado (DAG 1: scoring semanal)')

dags/scoring_dag.py generado (DAG 1: scoring semanal)


---
## 5.7 DAG 2: Monitoring + Retraining (frecuencia baja)

Este DAG se ejecuta **mensualmente**. Su funcion es vigilar si el modelo sigue siendo 
fiable y, si detecta cambios significativos en los datos, reentrenar automaticamente.

### Por que es necesario?

El modelo se entreno con datos de un periodo concreto. Con el tiempo:
- Raona puede empezar a vender productos nuevos que atraen otro perfil
- El mercado cambia y responden perfiles diferentes
- LinkedIn modifica su algoritmo, alterando tasas de respuesta

El modelo sigue scoreando, pero sus predicciones pueden dejar de ser fiables. 
Este DAG lo detecta antes de que sea un problema.

### Flujo

```
evaluate_drift ──> check_drift
                       |
               ┌──────┴──────┐
               |             |
         PSI > 0.25    PSI <= 0.25
               |             |
           retrain       skip (fin)
               |
           validate
               |
            deploy
```

| Paso | Modulo | Que hace |
|------|--------|----------|
| **evaluate_drift** | `pipelines/monitor.py` | Calcula PSI por feature comparando datos recientes vs entrenamiento |
| **check_drift** | DAG logic | Si PSI > 0.25 en 3+ features: reentrenar. Si no: fin |
| **retrain** | `pipelines/retrain.py` | Reentrena LightGBM con datos acumulados |
| **validate** | `pipelines/validate.py` | Compara nuevo modelo vs actual en holdout. Solo promueve si es mejor |
| **deploy** | `pipelines/validate.py` | Reemplaza `lead_scorer.pkl` en el Container 1 |

### Frecuencia

Mensual. Podrian pasar meses sin reentrenar si los datos son estables.

El siguiente codigo genera `app/dags/monitoring_dag.py`:

In [19]:
# Generar monitoring_dag.py (DAG 2: monitoring mensual)
monitoring_dag_code = '''"""Airflow DAG 2: Monitoring mensual + reentrenamiento condicional.

Pipeline: monitor >> check_drift >> [retrain >> validate | skip_retrain] >> done

Se ejecuta mensualmente. Si detecta drift significativo (PSI > 0.25 en 3+ features),
dispara reentrenamiento automatico y valida el modelo candidato.
"""
from datetime import datetime, timedelta
from airflow import DAG
from airflow.operators.python import PythonOperator, BranchPythonOperator
from airflow.operators.empty import EmptyOperator

import os
import sys
import pickle
import pandas as pd
import logging

logger = logging.getLogger(__name__)

BASE_DIR = os.environ.get("LEAD_SCORING_BASE_DIR", "/opt/airflow/data")
MODEL_DIR = os.environ.get("LEAD_SCORING_MODEL_DIR", "/opt/airflow/models")
PROCESSED_DIR = os.path.join(BASE_DIR, "processed")
SCORED_DIR = os.path.join(BASE_DIR, "scored")

PIPELINES_DIR = os.environ.get("PIPELINES_DIR", "/opt/airflow/dags")
if PIPELINES_DIR not in sys.path:
    sys.path.insert(0, PIPELINES_DIR)

default_args = {
    "owner": "raona-data",
    "depends_on_past": False,
    "email_on_failure": False,
    "retries": 1,
    "retry_delay": timedelta(minutes=5),
}


def _monitor(**kwargs):
    """Calcula PSI y decide si reentrenar."""
    from pipelines import monitor
    train_path = os.path.join(PROCESSED_DIR, "historical.parquet")
    new_path = os.path.join(PROCESSED_DIR, "transformed.parquet")
    feature_names_path = os.path.join(MODEL_DIR, "feature_names.pkl")
    with open(feature_names_path, "rb") as f:
        features = pickle.load(f)
    psi_df = monitor.run(train_path, new_path, features)
    psi_path = os.path.join(SCORED_DIR, f"psi_{datetime.now().strftime('%Y%m%d')}.csv")
    psi_df.to_csv(psi_path, index=False)
    n_alerts = (psi_df["status"] == "ALERTA").sum()
    kwargs["ti"].xcom_push(key="n_drift_alerts", value=n_alerts)
    return n_alerts


def _check_drift(**kwargs):
    """Branch: reentrenar solo si hay drift significativo."""
    n_alerts = kwargs["ti"].xcom_pull(task_ids="monitor", key="n_drift_alerts")
    if n_alerts and n_alerts > 0:
        return "retrain"
    return "skip_retrain"


def _retrain(**kwargs):
    """Reentrena modelo con datos acumulados."""
    from pipelines import retrain
    hist_path = os.path.join(PROCESSED_DIR, "historical.parquet")
    new_path = os.path.join(PROCESSED_DIR, "transformed.parquet")
    hist = pd.read_parquet(hist_path)
    new = pd.read_parquet(new_path)
    combined = pd.concat([hist, new], ignore_index=True)
    combined_path = os.path.join(PROCESSED_DIR, "combined_for_retrain.parquet")
    combined.to_parquet(combined_path, index=False)
    feature_names_path = os.path.join(MODEL_DIR, "feature_names.pkl")
    result = retrain.run(combined_path, MODEL_DIR, feature_names_path)
    kwargs["ti"].xcom_push(key="retrain_result", value=result)
    return result


def _validate(**kwargs):
    """Valida modelo candidato vs actual."""
    from pipelines import validate
    retrain_result = kwargs["ti"].xcom_pull(task_ids="retrain", key="retrain_result")
    if not retrain_result or retrain_result.get("status") != "success":
        logger.warning("Reentrenamiento fallido, saltando validacion")
        return
    holdout_path = os.path.join(PROCESSED_DIR, "transformed.parquet")
    feature_names_path = os.path.join(MODEL_DIR, "feature_names.pkl")
    result = validate.run(holdout_path, MODEL_DIR, retrain_result["candidate_dir"], feature_names_path)
    return result


with DAG(
    dag_id="raona_monitoring_monthly",
    default_args=default_args,
    description="Monitoring mensual: PSI drift -> reentrenamiento condicional",
    schedule_interval="@monthly",
    start_date=datetime(2026, 1, 1),
    catchup=False,
    tags=["raona", "ml", "monitoring"],
) as dag:

    monitor = PythonOperator(task_id="monitor", python_callable=_monitor)
    check_drift = BranchPythonOperator(task_id="check_drift", python_callable=_check_drift)
    retrain = PythonOperator(task_id="retrain", python_callable=_retrain)
    validate = PythonOperator(task_id="validate", python_callable=_validate)
    skip_retrain = EmptyOperator(task_id="skip_retrain")
    done = EmptyOperator(task_id="done", trigger_rule="none_failed_min_one_success")

    monitor >> check_drift
    check_drift >> retrain >> validate >> done
    check_drift >> skip_retrain >> done
'''

with open(os.path.join(DAGS_DIR, 'monitoring_dag.py'), 'w') as f:
    f.write(monitoring_dag_code)

# Eliminar el DAG unificado antiguo
old_dag = os.path.join(DAGS_DIR, 'lead_scoring_dag.py')
if os.path.exists(old_dag):
    os.remove(old_dag)
    print(f'Eliminado: lead_scoring_dag.py (reemplazado por 2 DAGs separados)')

print('dags/monitoring_dag.py generado (DAG 2: monitoring mensual)')
print(f'\n=== DAGs en {DAGS_DIR} ===')
for f_name in sorted(os.listdir(DAGS_DIR)):
    print(f'  {f_name}')

dags/monitoring_dag.py generado (DAG 2: monitoring mensual)

=== DAGs en /Users/acaballito/Library/CloudStorage/GoogleDrive-adriana.caballero@gmail.com/.shortcut-targets-by-id/1LzrxzfxIAZukyDLfOvioF7Z2ESoOXZGz/TFM/acaballero/TFM_deliverables/app/dags ===
  __pycache__
  monitoring_dag.py
  pipelines
  scoring_dag.py


### Implementación del PSI (Population Stability Index)

El PSI compara la distribución de cada feature entre los datos de entrenamiento 
y los datos más recientes:

| PSI | Interpretación | Acción |
|-----|---------------|--------|
| < 0.10 | Sin cambio significativo | Nada |
| 0.10 - 0.25 | Cambio moderado | Monitorizar |
| > 0.25 | Cambio significativo | Considerar reentrenamiento |

In [20]:
def calculate_psi(expected, actual, bins=10):
    """Calcula el Population Stability Index entre dos distribuciones."""
    breakpoints = np.percentile(expected[~np.isnan(expected)],
                                np.linspace(0, 100, bins + 1))
    breakpoints = np.unique(breakpoints)
    
    if len(breakpoints) < 3:
        return 0.0
    
    expected_counts = np.histogram(expected[~np.isnan(expected)], bins=breakpoints)[0]
    actual_counts = np.histogram(actual[~np.isnan(actual)], bins=breakpoints)[0]
    
    expected_pct = expected_counts / max(expected_counts.sum(), 1) + 1e-6
    actual_pct = actual_counts / max(actual_counts.sum(), 1) + 1e-6
    
    psi = np.sum((actual_pct - expected_pct) * np.log(actual_pct / expected_pct))
    return psi

print('Funcion calculate_psi() definida')

Funcion calculate_psi() definida


In [21]:
# Simular monitorizacion: dividir datos en referencia y produccion
df = pd.read_parquet(os.path.join(WORKING_DATA, 'predictions.parquet'))

n_half = len(df) // 2
df_train_ref = df.iloc[:n_half]
df_prod_sim = df.iloc[n_half:]

# Nota: como no disponemos de una columna temporal, dividimos el dataset por posicion.
# Esto equivale a un split aleatorio, no temporal. En produccion, el PSI se calcularia
# entre los datos de entrenamiento originales y los datos nuevos del ultimo mes.
print(f'Datos de referencia (entrenamiento): {len(df_train_ref):,} filas')
print(f'Datos simulados (produccion): {len(df_prod_sim):,} filas')

# Calcular PSI para cada feature del modelo
psi_results = []
for col in FEATURE_COLS:
    if col in df.columns:
        psi = calculate_psi(df_train_ref[col].values, df_prod_sim[col].values)
        status = 'OK' if psi < 0.1 else ('MONITORIZAR' if psi < 0.25 else 'ALERTA')
        psi_results.append({'Feature': col, 'PSI': psi, 'Status': status})

psi_df = pd.DataFrame(psi_results).sort_values('PSI', ascending=False)
print('\n=== PSI por feature ===')
print(psi_df.to_string(index=False))

Datos de referencia (entrenamiento): 2,993 filas
Datos simulados (produccion): 2,994 filas



=== PSI por feature ===
                    Feature      PSI      Status
           fe_log_employees 0.181015 MONITORIZAR
        Number of employees 0.181015 MONITORIZAR
           nlp_embedding_01 0.158765 MONITORIZAR
           nlp_embedding_02 0.149021 MONITORIZAR
     fe_company_size_bucket 0.146012 MONITORIZAR
      fe_headcount_momentum 0.123971 MONITORIZAR
     fe_type_of_contact_ord 0.116110 MONITORIZAR
 Two years headcount growth 0.109045 MONITORIZAR
Six months headcount growth 0.105107 MONITORIZAR
           nlp_embedding_03 0.099063          OK
      fe_department_encoded 0.077968          OK
    Yearly headcount growth 0.077795          OK
      ext_ms_maturity_score 0.072872          OK
               Year founded 0.062031          OK
  nlp_contact_report_length 0.056024          OK
             fe_company_age 0.053566          OK
                  nlp_topic 0.035623          OK
          nlp_urgency_score 0.029411          OK
          nlp_report_length 0.018129        

In [22]:
# Visualizar PSI
colors = ['#e74c3c' if s == 'ALERTA' else '#f39c12' if s == 'MONITORIZAR' else '#2ecc71'
          for s in psi_df['Status']]

fig = go.Figure()
fig.add_trace(go.Bar(
    x=psi_df['Feature'], y=psi_df['PSI'],
    marker_color=colors,
    text=psi_df['PSI'].round(3), textposition='auto'
))
fig.add_hline(y=0.1, line_dash='dash', line_color='#f39c12',
              annotation_text='Umbral monitorizar (0.1)')
fig.add_hline(y=0.25, line_dash='dash', line_color='#e74c3c',
              annotation_text='Umbral alerta (0.25)')
fig.update_layout(
    title='Population Stability Index (PSI) por feature',
    xaxis_title='Feature', yaxis_title='PSI',
    height=500, xaxis_tickangle=-45
)
fig.show()

In [23]:
# Simular drift artificial para demostrar la deteccion
print('=== Simulacion de drift artificial ===')
print('Escenario: Number of employees aumenta un 50% en datos nuevos\n')

df_drift = df_prod_sim.copy()
df_drift['Number of employees'] = df_drift['Number of employees'] * 1.5

psi_normal = calculate_psi(
    df_train_ref['Number of employees'].values,
    df_prod_sim['Number of employees'].values
)
psi_drift = calculate_psi(
    df_train_ref['Number of employees'].values,
    df_drift['Number of employees'].values
)

print(f'PSI sin drift:  {psi_normal:.4f} -> {"OK" if psi_normal < 0.1 else "ALERTA"}')
print(f'PSI con drift:  {psi_drift:.4f} -> {"OK" if psi_drift < 0.1 else "ALERTA"}')
print(f'\nEl DAG 2 detectaria este cambio y dispararia el reentrenamiento.')

=== Simulacion de drift artificial ===
Escenario: Number of employees aumenta un 50% en datos nuevos



PSI sin drift:  0.1810 -> ALERTA
PSI con drift:  0.3891 -> ALERTA

El DAG 2 detectaria este cambio y dispararia el reentrenamiento.


---
## 5.8 Arquitectura completa

> **Nota:** El diagrama siguiente muestra la arquitectura de **produccion propuesta**. 
> La **demo desplegada** (raona-lead-scoring.streamlit.app) usa Streamlit Cloud con los 
> mismos artefactos `.pkl` pero sin Airflow ni Docker.

### Como encajan las piezas

```
Enginy (exporta CSV nuevos)
         |
    DAG 1: Scoring (semanal)  ──── scoring_dag.py
         |
         ├── ingest > transform > score > deliver
         |                                  |
         |                         CSV rankeado para
         |                         equipo comercial
         |
    DAG 2: Monitoring (mensual) ── monitoring_dag.py
         |
         ├── evaluate_drift > check
         |                     |
         |              si drift: retrain > validate > deploy
         |                                              |
         |                                    actualiza .pkl
         |                                              |
    Container 1: API (siempre activo)  <────────────────┘
         |
         ├── POST /score (scoring individual)
         ├── Streamlit conecta aqui
         └── CRM / integraciones
```

### Estructura de archivos

```
TFM_deliverables/
├── app/
│   ├── api.py                        # FastAPI (Container API)
│   ├── streamlit_app.py              # App interactiva (demo, 3 tabs)
│   ├── Dockerfile                    # Container API: Python + FastAPI
│   ├── Dockerfile.airflow            # Container Airflow: + libgomp1 + ML deps
│   ├── docker-compose.yml            # Orquestacion: 6 servicios
│   ├── requirements-api.txt          # Dependencias API (produccion Docker)
│   ├── requirements.txt              # Dependencias Streamlit (demo)
│   ├── models/
│   │   ├── lead_scorer.pkl           # Modelo de PRODUCCION
│   │   ├── preprocessor.pkl
│   │   ├── clustering.pkl
│   │   └── feature_names.pkl
│   ├── dags/
│   │   ├── scoring_dag.py            # DAG 1: Scoring semanal
│   │   └── monitoring_dag.py         # DAG 2: Monitoring mensual
│   └── pipelines/
│       ├── __init__.py
│       ├── ingest.py                 # NB01: carga y validacion
│       ├── transform.py              # NB03: feature engineering (39 features)
│       ├── score.py                  # NB04: scoring + clustering
│       ├── monitor.py                # NB05: PSI drift detection
│       ├── retrain.py                # NB05: reentrenamiento
│       └── validate.py               # NB05: validacion y promocion
├── notebooks/
│   ├── 01_data_loading.ipynb
│   ├── 02_eda.ipynb
│   ├── 03_feature_engineering.ipynb
│   ├── 04_models.ipynb
│   └── 05_mlops.ipynb
├── report/
│   ├── index.html                    # Data Story
│   ├── nb01-nb05 HTML exports
│   └── charts/
└── _working/
    ├── data/
    └── mlruns/
```

---
## 5.9 Trabajo futuro y roadmap de produccion

### Mejoras del modelo

| Mejora | Justificacion | Esfuerzo |
|--------|--------------|----------|
| Datos externos INE (sector/digitalizacion) | Enriquecer features ext_ con datos publicos del sector para mejorar el perfilado de empresas | Medio |
| Datos de engagement previo | Si Raona tiene historico de interacciones previas con una empresa (reuniones, propuestas), incorporarlo como feature de relacion previa mejoraria la prediccion | Medio |

### Integracion con Enginy

1. **Exportacion programatica**: script que descarga nuevos contactos via API de Enginy
2. **Scoring automatico**: DAG 1 se dispara al detectar nuevo CSV
3. **Priorizacion en campana**: ordenar contactos por lead_score antes del outreach
4. **Feedback loop**: resultados del outreach retroalimentan el reentrenamiento

### A/B Testing

Para validar el impacto real del modelo:
- **Grupo A**: Outreach priorizado por lead_score (top 50%)
- **Grupo B**: Outreach aleatorio (baseline actual de Raona)
- **Metrica**: reply rate del grupo A vs grupo B
- **Duracion**: 4-6 semanas, 500 contactos por grupo
- **Hipotesis**: basandonos en el lift@10% de 2.6x, esperamos que el grupo A tenga una tasa de respuesta significativamente superior

---
## Resumen

### Artefactos generados desde este notebook

| Archivo | Descripcion |
|---------|------------|
| `app/api.py` | Endpoint FastAPI (POST /score, GET /health) |
| `app/Dockerfile` | Container API: Python + FastAPI |
| `app/Dockerfile.airflow` | Container Airflow: imagen oficial + libgomp1 + ML deps |
| `app/docker-compose.yml` | Orquestacion: 6 servicios (API + Postgres + 4 Airflow) |
| `app/requirements-api.txt` | Dependencias del API (produccion Docker) |
| `app/pipelines/ingest.py` | Pipeline de ingestion (NB01) |
| `app/pipelines/transform.py` | Pipeline de feature engineering (NB03, 39 features) |
| `app/pipelines/score.py` | Pipeline de scoring (NB04) |
| `app/pipelines/monitor.py` | Pipeline de monitoring PSI (NB05) |
| `app/pipelines/retrain.py` | Pipeline de reentrenamiento (NB05) |
| `app/pipelines/validate.py` | Pipeline de validacion (NB05) |
| `app/dags/scoring_dag.py` | DAG 1: Scoring semanal |
| `app/dags/monitoring_dag.py` | DAG 2: Monitoring mensual |

### Artefactos preexistentes (no generados aqui)

| Archivo | Descripcion |
|---------|------------|
| `app/streamlit_app.py` | Demo interactiva (Streamlit Cloud) |
| `app/requirements.txt` | Dependencias Streamlit (demo) |
| `models/lead_scorer.pkl` | Modelo de PRODUCCION (39 features) |
| `models/preprocessor.pkl` | Pipeline preprocesado |
| `models/clustering.pkl` | K-Means + scaler |
| `models/feature_names.pkl` | Lista de features |

### Arquitectura de produccion

| Componente | Funcion | Frecuencia |
|-----------|---------|------------|
| **Container API** | Scoring individual via REST (FastAPI) | Siempre activo |
| **Airflow** (4 servicios + Postgres) | Orquestacion de pipelines | Siempre activo |
| **DAG 1 (Scoring)** | Batch scoring de contactos nuevos | Semanal |
| **DAG 2 (Monitoring)** | Vigilancia de drift + reentrenamiento | Mensual |
| **Streamlit** | Demo interactiva para el equipo comercial | Siempre activo |

### Modelo en produccion

| Metrica | Valor |
|---------|-------|
| PR-AUC | 0.303 |
| Precision@100 | 39% (vs 14% aleatorio) |
| Lift@10% | 2.6x |
| Features | 39 |

### Conclusion

El sistema va mas alla de un modelo aislado: es un **pipeline de produccion profesional** 
con scoring en tiempo real (API), procesamiento en batch (DAG 1), monitorizacion continua 
(DAG 2) y una interfaz accesible (Streamlit).

La infraestructura se despliega en **dos bloques Docker** (API + Airflow) con seis servicios 
orquestados por `docker-compose.yml`. Cada bloque tiene su propio Dockerfile con dependencias 
especificas, siguiendo el principio de responsabilidad unica.

La **demo en Streamlit Cloud** (raona-lead-scoring.streamlit.app) permite al equipo comercial 
explorar los resultados hoy. La **arquitectura de produccion** (API + Airflow + Docker) 
esta lista para desplegarse cuando Raona decida escalar el sistema.

In [24]:
# Inventario final
print('=' * 60)
print('RESUMEN NOTEBOOK 05 - MLOps')
print('=' * 60)

generated_files = [
    # Infraestructura
    ('app/api.py', 'FastAPI endpoint'),
    ('app/Dockerfile', 'Container API: Python + FastAPI'),
    ('app/Dockerfile.airflow', 'Container Airflow: + libgomp1 + ML deps'),
    ('app/docker-compose.yml', 'Orquestacion: 6 servicios'),
    ('app/requirements-api.txt', 'Dependencias API (produccion Docker)'),
    # Pipelines
    ('app/pipelines/__init__.py', 'Modulo pipelines'),
    ('app/pipelines/ingest.py', 'NB01: ingestion'),
    ('app/pipelines/transform.py', 'NB03: feature engineering (39 features)'),
    ('app/pipelines/score.py', 'NB04: scoring + clustering'),
    ('app/pipelines/monitor.py', 'NB05: PSI drift detection'),
    ('app/pipelines/retrain.py', 'NB05: reentrenamiento'),
    ('app/pipelines/validate.py', 'NB05: validacion'),
    # DAGs
    ('app/dags/scoring_dag.py', 'DAG 1: scoring semanal'),
    ('app/dags/monitoring_dag.py', 'DAG 2: monitoring mensual'),
    # Modelo (preexistente)
    ('app/models/lead_scorer.pkl', 'Modelo (PRODUCCION)'),
    ('app/models/preprocessor.pkl', 'Pipeline preprocesado'),
    ('app/models/clustering.pkl', 'K-Means + scaler'),
    ('app/models/feature_names.pkl', 'Lista de features'),
    # Demo (preexistente)
    ('app/streamlit_app.py', 'Demo Streamlit Cloud'),
    ('app/requirements.txt', 'Dependencias Streamlit (demo)'),
]

print('\nArchivos del sistema:')
for path, desc in generated_files:
    full_path = os.path.join(PROJECT_ROOT, path)
    exists = 'OK' if os.path.exists(full_path) else 'PENDIENTE'
    print(f'  [{exists}] {path} -- {desc}')

# Verificar que no queda el DAG antiguo
old_dag = os.path.join(PROJECT_ROOT, 'app', 'dags', 'lead_scoring_dag.py')
if os.path.exists(old_dag):
    print(f'\n  [AVISO] lead_scoring_dag.py todavia existe (deberia haberse eliminado)')
else:
    print(f'\n  lead_scoring_dag.py eliminado correctamente (reemplazado por 2 DAGs)')

print(f'\nModelo en produccion: lead_scorer.pkl ({len(FEATURE_COLS)} features)')
print(f'Infraestructura: 2 bloques Docker, 6 servicios')

RESUMEN NOTEBOOK 05 - MLOps

Archivos del sistema:
  [OK] app/api.py -- FastAPI endpoint
  [OK] app/Dockerfile -- Container API: Python + FastAPI
  [OK] app/Dockerfile.airflow -- Container Airflow: + libgomp1 + ML deps
  [OK] app/docker-compose.yml -- Orquestacion: 6 servicios
  [OK] app/requirements-api.txt -- Dependencias API (produccion Docker)
  [OK] app/pipelines/__init__.py -- Modulo pipelines
  [OK] app/pipelines/ingest.py -- NB01: ingestion
  [OK] app/pipelines/transform.py -- NB03: feature engineering (39 features)
  [OK] app/pipelines/score.py -- NB04: scoring + clustering
  [OK] app/pipelines/monitor.py -- NB05: PSI drift detection
  [OK] app/pipelines/retrain.py -- NB05: reentrenamiento
  [OK] app/pipelines/validate.py -- NB05: validacion
  [OK] app/dags/scoring_dag.py -- DAG 1: scoring semanal
  [OK] app/dags/monitoring_dag.py -- DAG 2: monitoring mensual
  [OK] app/models/lead_scorer.pkl -- Modelo (PRODUCCION)
  [OK] app/models/preprocessor.pkl -- Pipeline preprocesado
  